### PR TITLE
feat(rule): add blaze-consistent-eventmap-parameters

### LINKED ISSUE #12: Rule Proposal: blaze-consistent-eventmap-params
Consisten argument names for event maps.

```js
Template.x.events({
  'click #button': function (event, templateInstance) {}
})
```

Ensures that `event` and `templateInstance` are always called exactly that.

### PR DESCRIPTION
**Pull request by PhilippSpo**
Add rule that enforces consistent event handler parameters.

closes #12

### CODE REVIEW

File: lib/rules/blaze-consistent-eventmap-parameters.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+  // ---------------------------------------------------------------------------
+  // Public
+  // ---------------------------------------------------------------------------
+
+  return {
+
+    CallExpression(node) {
+
+      // const config = context.options[0]
```
Comment (dferber90) [inline lib/rules/blaze-consistent-eventmap-parameters.js]: Remove this?

Comment (PhilippSpo) [inline lib/rules/blaze-consistent-eventmap-parameters.js]: dome

File: lib/rules/blaze-consistent-eventmap-parameters.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import {CLIENT, UNIVERSAL} from '../util/environment'
+import {isInServerBlock, isFunction} from '../util/ast'
+
+// -----------------------------------------------------------------------------
+// Rule Definition
+// -----------------------------------------------------------------------------
+
+module.exports = getMeta => context => {
+
+  const {env} = getMeta(context.getFilename())
```
Comment (dferber90) [inline lib/rules/blaze-consistent-eventmap-parameters.js]: const {env, isLintedEnv} = getMeta(..)

File: lib/rules/blaze-consistent-eventmap-parameters.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+  if (env !== CLIENT && env !== UNIVERSAL) {
```
Comment (dferber90) [inline lib/rules/blaze-consistent-eventmap-parameters.js]: `if (!isLintedEnv) { return {} }`

File: lib/rules/blaze-consistent-eventmap-parameters.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    const param = node.params[paramIndex]
+    if (param && param.name !== expectedParamName) {
+      context.report(
+        param,
+        `Invalid parameter name, use "${expectedParamName}" instead`
+      )
+    }
+  }
+
+  function validateEventDef (eventDefNode, ancestors) {
```
Comment (dferber90) [inline lib/rules/blaze-consistent-eventmap-parameters.js]: `function validateEventDef (eventDefNode, executors) {`

File: lib/rules/blaze-consistent-eventmap-parameters.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    if (env === UNIVERSAL && isInServerBlock(ancestors)) {
```
Comment (dferber90) [inline lib/rules/blaze-consistent-eventmap-parameters.js]: if (executors.has('client') || executors.has('cordova')) {

File: lib/rules/blaze-consistent-eventmap-parameters.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  // ---------------------------------------------------------------------------
+  // Public
+  // ---------------------------------------------------------------------------
+
+  return {
+
+    CallExpression(node) {
+
+      const ancestors = context.getAncestors()
+      // check if the Template.x.events(/*eventMap*/) call got any arguments
```
Comment (dferber90) [inline lib/rules/blaze-consistent-eventmap-parameters.js]: `const executors = getExecutors(env, ancestors)`

File: lib/rules/blaze-consistent-eventmap-parameters.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      if (node.arguments.length > 0) {
+        const eventMap = node.arguments[0]
+        // check if the event map argument is an object
+        if (
+          eventMap.type === 'ObjectExpression'
+        ) {
+          eventMap.properties.forEach((eventDef) => validateEventDef(eventDef, ancestors))
```
Comment (dferber90) [inline lib/rules/blaze-consistent-eventmap-parameters.js]: `eventMap.properties.forEach((eventDef) => validateEventDef(eventDef, executors))`

File: tests/lib/rules/blaze-consistent-eventmap-parameters.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+} from '../../../dist/util/environment.js'
+const rule = require('../../../dist/rules/blaze-consistent-eventmap-parameters')
+const RuleTester = require('eslint').RuleTester
+
+// -----------------------------------------------------------------------------
+// Tests
+// -----------------------------------------------------------------------------
+
+const ruleTester = new RuleTester()
+ruleTester.run('blaze-consistent-eventmap-parameters', rule(() => ({env: CLIENT})), {
```
Comment (dferber90) [inline tests/lib/rules/blaze-consistent-eventmap-parameters.js]: `{env: CLIENT, isLintedEnv: true}`

File: tests/lib/rules/blaze-consistent-eventmap-parameters.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      `,
+      errors: [
+        {message: 'Allowed on client only', type: 'Property'}
+      ]
+    }
+  ]
+
+})
+
+ruleTester.run('pubsub', rule(() => ({env: NON_METEOR})), {
```
Comment (dferber90) [inline tests/lib/rules/blaze-consistent-eventmap-parameters.js]: `rule(() => ({isLintedEnv: false})`

File: lib/rules/blaze-consistent-eventmap-parameters.js (hunk 5 above)
Comment (PhilippSpo) [inline lib/rules/blaze-consistent-eventmap-parameters.js]: Should it not be:
```js
if (executors.has('server')) {
      return context.report(
        eventDefNode,
        `Allowed on client only`
      )
    }
```
Or is this checked in another rule?

**Comment by dferber90** [comment]
Closed through be73ffa031a9d7542d2e33dc658483dd50e49e54.

**Pull request closed without merging**